File: matlab_winder/WinderJob.m
classdef WinderJob < handle
    properties
        machine
        mandrel
        layers = Layer.empty
    end

    methods
        function obj = WinderJob(machine, mandrel)
            obj.machine = machine;
            obj.mandrel = mandrel;
        end

        function addLayer(obj, angle, passes)
            obj.layers(end+1) = Layer(angle, passes, "helical");
        end

        function addHoopLayer(obj, passes)
            obj.layers(end+1) = Layer(90, passes, "hoop");
        end

        function gcode = generateGCode(obj)
            gcode = strings(0,1);

            if obj.machine.units == "in"
                gcode(end+1,1) = "G20 ; units in inches";
            else
                gcode(end+1,1) = "G21 ; units in mm";
            end

            gcode(end+1,1) = "G90 ; absolute positioning";
            gcode(end+1,1) = "G92 " + obj.machine.linAxis + "0 " + obj.machine.rotAxis + "0 ; set current position as zero";
            gcode(end+1,1) = "G0 " + obj.machine.linAxis + "0 " + obj.machine.rotAxis + "0";

            currentRot = 0; % mandrel revs
            currentLin = 0; % carriage position

            for i = 1:length(obj.layers)
                layer = obj.layers(i);

                if layer.type == "helical"
                    [newLines, currentRot, currentLin] = obj.generateHelicalLayer(layer, currentRot, currentLin);
                    gcode = [gcode; newLines(:)];

                elseif layer.type == "hoop"
                    [newLines, currentRot, currentLin] = obj.generateHoopLayer(layer, currentRot, currentLin);
                    gcode = [gcode; newLines(:)];
                end
            end

            gcode(end+1,1) = "M30 ; end program";
        end

        function [lines, currentRot, currentLin] = generateHelicalLayer(obj, layer, currentRot, currentLin)
            lines = strings(0,1);

            theta = deg2rad(abs(layer.angle));
            C = obj.mandrel.circumference();
            L = obj.mandrel.length;

            % axial travel per mandrel revolution
            lead = C / tan(theta);

            % mandrel revolutions needed for one full carriage pass
            deltaRot = L / lead;

            for p = 1:layer.passes
                % Alternate carriage endpoint each pass
                if mod(p,2) == 1
                    nextLin = L;
                else
                    nextLin = 0;
                end

                % Positive/negative angle controls mandrel rotation direction
                currentRot = currentRot + deltaRot;

                currentLin = nextLin;

                cmdRot = currentRot * obj.machine.rotScale;
                cmdLin = currentLin * obj.machine.linScale;

                lines(end+1,1) = ...
                    "G1 " + obj.machine.rotAxis + num2str(cmdRot, "%.4f") + ...
                    " " + obj.machine.linAxis + num2str(cmdLin, "%.4f") + ...
                    " F" + num2str(obj.machine.feedRate);
            end
        end

        function [lines, currentRot, currentLin] = generateHoopLayer(obj, layer, currentRot, currentLin)
            lines = strings(0,1);

            for p = 1:layer.passes
                currentRot = currentRot + 1; % one mandrel revolution

                cmdRot = currentRot * obj.machine.rotScale;
                cmdLin = currentLin * obj.machine.linScale;

                lines(end+1,1) = ...
                    "G1 " + obj.machine.rotAxis + num2str(cmdRot, "%.4f") + ...
                    " " + obj.machine.linAxis + num2str(cmdLin, "%.4f") + ...
                    " F" + num2str(obj.machine.feedRate);
            end
        end
    end
end

File: python_winder/winder_job.py
import math
from machine import Machine
from mandrel import Mandrel
from layer import Layer


class WinderJob:
    def __init__(self, machine: Machine, mandrel: Mandrel):
        self.machine = machine
        self.mandrel = mandrel
        self.layers: list[Layer] = []

    def add_layer(self, angle: float, passes: int) -> None:
        self.layers.append(Layer(angle, passes, "helical"))

    def add_hoop_layer(self, passes: int) -> None:
        self.layers.append(Layer(90, passes, "hoop"))

    def generate_gcode(self) -> list[str]:
        gcode = []

        if self.machine.units == "in":
            gcode.append("G20 ; units in inches")
        else:
            gcode.append("G21 ; units in mm")

        gcode.append("G90 ; absolute positioning")
        gcode.append(
            f"G92 {self.machine.lin_axis}0 {self.machine.rot_axis}0 ; set current position as zero"
        )
        gcode.append(
            f"G0 {self.machine.lin_axis}0 {self.machine.rot_axis}0"
        )

        current_rot = 0.0  # mandrel revolutions
        current_lin = 0.0  # carriage position

        for layer in self.layers:
            if layer.type == "helical":
                new_lines, current_rot, current_lin = self._generate_helical_layer(
                    layer, current_rot, current_lin
                )
            elif layer.type == "hoop":
                new_lines, current_rot, current_lin = self._generate_hoop_layer(
                    layer, current_rot, current_lin
                )
            else:
                continue
            gcode.extend(new_lines)

        gcode.append("M30 ; end program")
        return gcode

    def _generate_helical_layer(
        self, layer: Layer, current_rot: float, current_lin: float
    ) -> tuple[list[str], float, float]:
        lines = []

        theta = math.radians(abs(layer.angle))
        C = self.mandrel.circumference()
        L = self.mandrel.length

        # axial travel per mandrel revolution
        lead = C / math.tan(theta)

        # mandrel revolutions needed for one full carriage pass
        delta_rot = L / lead

        for p in range(1, layer.passes + 1):
            # alternate carriage endpoint each pass
            next_lin = L if p % 2 == 1 else 0.0

            current_rot += delta_rot
            current_lin = next_lin

            cmd_rot = current_rot * self.machine.rot_scale
            cmd_lin = current_lin * self.machine.lin_scale

            lines.append(
                f"G1 {self.machine.rot_axis}{cmd_rot:.4f}"
                f" {self.machine.lin_axis}{cmd_lin:.4f}"
                f" F{self.machine.feed_rate}"
            )

        return lines, current_rot, current_lin

    def _generate_hoop_layer(
        self, layer: Layer, current_rot: float, current_lin: float
    ) -> tuple[list[str], float, float]:
        lines = []

        for _ in range(layer.passes):
            current_rot += 1.0  # one mandrel revolution

            cmd_rot = current_rot * self.machine.rot_scale
            cmd_lin = current_lin * self.machine.lin_scale

            lines.append(
                f"G1 {self.machine.rot_axis}{cmd_rot:.4f}"
                f" {self.machine.lin_axis}{cmd_lin:.4f}"
                f" F{self.machine.feed_rate}"
            )

        return lines, current_rot, current_lin
